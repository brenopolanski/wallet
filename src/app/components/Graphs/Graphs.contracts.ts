import React, { MouseEvent, MutableRefObject } from "react";

type GraphType = "line" | "donut";

interface GraphDataPoint {
	color: string;
	data: Record<string, any>;
	value: number;
}

const GRAPH_COLORS = ["success-600", "warning-600", "info-600", "danger-400", "hint-400", "secondary-400"];
const GRAPH_COLORS_DARK = ["success-600", "warning-600", "info-600", "danger-400", "hint-400", "secondary-600"];

const GRAPH_COLOR_EMPTY = "secondary-300";
const GRAPH_COLOR_EMPTY_DARK = "secondary-800";

const MIN_VALUE = 1;
const MIN_DISPLAY_VALUE_LINE = 2.5;
const MIN_DISPLAY_VALUE_DONUT = 5;

type UseGraphDataHook = (graphType: GraphType) => {
	normalizeData: (data: GraphDataPoint[]) => GraphDataPoint[];
};

type UseGraphWidthHook = () => [MutableRefObject<SVGSVGElement | null>, number];

type UseGraphTooltipHook = (
	renderFunction: ((dataPoint: GraphDataPoint) => JSX.Element) | undefined,
	type: GraphType,
) => {
	Tooltip: React.VFC;
	getMouseEventProperties: (dataPoint: GraphDataPoint) => {
		onMouseMove: (event: MouseEvent<SVGElement>) => void;
		onMouseOut: (event: MouseEvent<SVGElement>) => void;
	};
};

export type { GraphDataPoint, GraphType, UseGraphDataHook, UseGraphTooltipHook, UseGraphWidthHook };

export {
	GRAPH_COLOR_EMPTY,
	GRAPH_COLOR_EMPTY_DARK,
	GRAPH_COLORS,
	GRAPH_COLORS_DARK,
	MIN_DISPLAY_VALUE_DONUT,
	MIN_DISPLAY_VALUE_LINE,
	MIN_VALUE,
};
