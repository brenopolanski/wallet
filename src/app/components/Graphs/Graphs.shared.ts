import { MutableRefObject, useEffect, useState } from "react";

export const GRAPH_COLORS = ["primary-600", "warning-600", "info-600", "danger-400", "hint-400", "secondary-400"];

export const useGraphWidth = (svgReference: MutableRefObject<SVGSVGElement | null>): number => {
	const [value, setValue] = useState(0);

	useEffect(() => {
		const setWidth = () => {
			setValue(svgReference.current?.clientWidth ?? 0);
		};

		setWidth();

		window.addEventListener("resize", setWidth);

		return () => {
			window.removeEventListener("resize", setWidth);
		};
	});

	return value;
};
