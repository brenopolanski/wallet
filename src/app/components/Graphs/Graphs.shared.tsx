import React, { MouseEvent, MutableRefObject, useEffect, useRef, useState } from "react";

const GRAPH_COLORS = ["success-600", "warning-600", "info-600", "danger-400", "hint-400", "secondary-400"];

const useGraphWidth = (reference: MutableRefObject<SVGSVGElement | null>): number => {
	const [value, setValue] = useState(0);

	useEffect(() => {
		const setWidth = () => {
			setValue(reference.current?.clientWidth ?? 0);
		};

		setWidth();

		window.addEventListener("resize", setWidth);

		return () => {
			window.removeEventListener("resize", setWidth);
		};
	});

	return value;
};

const useGraphFormatter = (): { formatPercent: (value: number) => string } => ({
	formatPercent: (value: number): string => `${Math.round((value + Number.EPSILON) * 100) / 100}%`,
});

interface UseGraphTooltipReturnType<TDataPoint, TSvgElement> {
	Tooltip: React.VFC;
	getMouseEventProperties: (
		dataPoint: TDataPoint,
	) => {
		onMouseMove: (event: MouseEvent<TSvgElement>) => void;
		onMouseOut: (event: MouseEvent<TSvgElement>) => void;
	};
}

function useGraphTooltip<TDataPoint, TSvgElement>(
	renderFunction?: (dataPoint: TDataPoint) => JSX.Element,
): UseGraphTooltipReturnType<TDataPoint, TSvgElement> {
	const tooltipReference = useRef<HTMLDivElement | null>(null);

	const [tooltipDataPoint, setTooltipDataPoint] = useState<TDataPoint | undefined>(undefined);

	const Tooltip: React.VFC = () => (
		<div ref={tooltipReference} className="absolute hidden">
			{!!tooltipDataPoint && renderFunction?.(tooltipDataPoint)}
		</div>
	);

	const getMouseEventProperties = (dataPoint: TDataPoint) => ({
		onMouseMove: (event: MouseEvent<TSvgElement>) => {
			const tooltipElement = tooltipReference.current;

			if (!tooltipElement) {
				return;
			}

			if (!tooltipDataPoint) {
				setTooltipDataPoint(dataPoint);
			}

			tooltipElement.style.left = `${event.pageX - Math.floor(tooltipElement.clientWidth / 2)}px`;
			tooltipElement.style.top = `${event.pageY - 48}px`;

			tooltipElement.classList.remove("hidden");
		},
		onMouseOut: () => {
			setTooltipDataPoint(undefined);
			tooltipReference.current?.classList.add("hidden");
		},
	});

	return { Tooltip, getMouseEventProperties };
}

export { GRAPH_COLORS, useGraphFormatter, useGraphTooltip, useGraphWidth };
