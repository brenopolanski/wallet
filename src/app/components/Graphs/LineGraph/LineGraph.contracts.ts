export interface LineGraphDataItem {
	label: string;
	value: number;
}

export type LineGraphData = LineGraphDataItem[];

export interface LineGraphProperties {
	data: LineGraphData;
}
