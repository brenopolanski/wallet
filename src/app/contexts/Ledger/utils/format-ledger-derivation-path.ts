import { LedgerDerivationScheme } from "app/contexts/Ledger/contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
