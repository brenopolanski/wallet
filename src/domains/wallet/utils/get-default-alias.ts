import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";

interface GetDefaultAliasInput {
	profile: Contracts.IProfile;
	network: Networks.Network;
}

export const getDefaultAlias = ({ profile, network }: GetDefaultAliasInput): string => {
	const makeAlias = (count: number) => `${network.displayName()} #${count}`;

	const sameCoinWallets = profile.wallets().findByCoinWithNetwork(network.coin(), network.id());

	let sameCoinWalletsCount = sameCoinWallets.length || 1;

	while (profile.wallets().findByAlias(makeAlias(sameCoinWalletsCount))) {
		sameCoinWalletsCount++;
	}

	return makeAlias(sameCoinWalletsCount);
};
