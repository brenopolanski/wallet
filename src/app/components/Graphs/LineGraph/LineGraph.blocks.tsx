import { useGraphFormatter } from "app/components/Graphs/Graphs.shared";
import React from "react";

import {
	LineGraphAnimationProperties,
	LineGraphLegendProperties,
	LineGraphSegmentProperties,
} from "./LineGraph.contracts";

const LineGraphAnimation: React.VFC<LineGraphAnimationProperties> = ({ animations }) => (
	<>
		{animations.map(({ attribute, from, to }, index) => (
			<React.Fragment key={index}>
				<animate attributeName={attribute} from={from} to={to} dur="0.1s" begin="mouseover" fill="freeze" />
				<animate attributeName={attribute} from={to} to={from} dur="0.15s" begin="mouseleave" fill="freeze" />
			</React.Fragment>
		))}
	</>
);

export const LineGraphSegment: React.VFC<LineGraphSegmentProperties> = ({
	config: { segmentHeight, segmentHeightHover },
	dataPoint: { x, width, color },
	onMouseMove,
	onMouseOut,
}) => (
	<rect
		onMouseMove={onMouseMove}
		onMouseOut={onMouseOut}
		x={x}
		y={segmentHeight}
		className={`fill-current text-theme-${color}`}
		width={width}
		height={segmentHeight}
		rx={segmentHeight / 2}
	>
		<LineGraphAnimation
			animations={[
				{ attribute: "height", from: segmentHeight, to: segmentHeightHover },
				{ attribute: "rx", from: segmentHeight / 2, to: segmentHeightHover / 2 },
				{ attribute: "y", from: segmentHeight, to: 0 },
			]}
		/>
	</rect>
);

export const LineGraphLegend: React.VFC<LineGraphLegendProperties> = ({ dataPoints }) => {
	const { formatPercent } = useGraphFormatter();

	return (
		<div className="flex space-x-4">
			{dataPoints.map(({ color, label, percent }, index) => (
				<div className="flex items-center space-x-1" key={index}>
					<div className={`h-3 w-1 rounded bg-theme-${color}`} />
					<div className="text-sm font-semibold text-theme-secondary-700 dark:text-theme-secondary-200">
						{label}
					</div>
					<div className="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700">
						{formatPercent(percent)}
					</div>
				</div>
			))}
		</div>
	);
};
