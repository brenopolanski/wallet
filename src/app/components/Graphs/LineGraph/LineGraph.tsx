import React from "react";

import { LineGraphProperties } from "./LineGraph.contracts";

export const LineGraph: React.VFC<LineGraphProperties> = ({ data }) => {
	console.log(data);

	return <></>;
};
