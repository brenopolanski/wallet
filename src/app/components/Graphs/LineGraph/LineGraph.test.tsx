import userEvent from "@testing-library/user-event";
import * as sharedGraphUtils from "app/components/Graphs/Graphs.shared";
import React from "react";
import { render, screen, waitFor } from "utils/testing-library";

import { LineGraph } from "./LineGraph";
import { LineGraphDataPoint } from "./LineGraph.contracts";

describe("LineGraph", () => {
	let data: LineGraphDataPoint[];
	let useWidthMock: jest.SpyInstance;

	beforeAll(() => {
		data = [
			{ color: "success-600", data: { label: "item 1" }, value: 50 },
			{ color: "warning-600", data: { label: "item 2" }, value: 30 },
			{ color: "info-600", data: { label: "item 3" }, value: 20 },
		];

		// Mock graph width to 100.
		useWidthMock = jest.spyOn(sharedGraphUtils, "useGraphWidth").mockReturnValue([undefined as never, 100]);
	});

	afterAll(() => {
		useWidthMock.mockRestore();
	});

	it("should render", () => {
		const { asFragment } = render(<LineGraph data={data} />);

		expect(screen.getByTestId("LineGraph__svg")).toBeInTheDocument();
		expect(screen.getAllByTestId("LineGraph__item")).toHaveLength(3);

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render as empty", () => {
		const { asFragment } = render(<LineGraph data={data} renderAsEmpty />);

		expect(screen.getByTestId("LineGraph__svg")).toBeInTheDocument();
		expect(screen.queryByTestId("LineGraph__item")).not.toBeInTheDocument();
		expect(screen.getByTestId("LineGraph__empty")).toBeInTheDocument();

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render tooltip on hover", async () => {
		const { asFragment } = render(
			<LineGraph
				data={data}
				renderTooltip={(dataPoint) => (
					<div data-testid="TooltipContent">
						{dataPoint.data.label} value: {dataPoint.value}
					</div>
				)}
			/>,
		);

		expect(screen.getByTestId("LineGraph__svg")).toBeInTheDocument();
		expect(screen.getAllByTestId("LineGraph__item")).toHaveLength(3);

		expect(screen.queryByTestId("TooltipContent")).not.toBeInTheDocument();

		userEvent.hover(screen.getAllByTestId("LineGraph__item")[0]);

		expect(screen.getByTestId("TooltipContent")).toBeInTheDocument();
		expect(screen.getByTestId("TooltipContent")).toHaveTextContent("item 1 value: 50");

		userEvent.unhover(screen.getAllByTestId("LineGraph__item")[0]);
		userEvent.hover(screen.getAllByTestId("LineGraph__item")[1]);

		expect(screen.getByTestId("TooltipContent")).toHaveTextContent("item 2 value: 30");

		userEvent.unhover(screen.getAllByTestId("LineGraph__item")[1]);

		await waitFor(() => expect(screen.getByTestId("TooltipContent").parentElement).toHaveClass("hidden"));

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render legend", () => {
		const { asFragment } = render(
			<LineGraph
				data={data}
				renderLegend={(dataPoints) => (
					<ul>
						{dataPoints.map((dataPoint, index) => (
							<li key={index}>
								{dataPoint.data.label} value: {dataPoint.value}
							</li>
						))}
					</ul>
				)}
			/>,
		);

		expect(screen.getByTestId("LineGraph__svg")).toBeInTheDocument();
		expect(screen.getAllByTestId("LineGraph__item")).toHaveLength(3);

		expect(screen.getByTestId("LineGraph__legend")).toBeInTheDocument();
		expect(screen.getAllByRole("listitem")).toHaveLength(3);
		expect(screen.getAllByRole("listitem")[0]).toHaveTextContent("item 1 value: 50");

		expect(asFragment()).toMatchSnapshot();
	});
});
