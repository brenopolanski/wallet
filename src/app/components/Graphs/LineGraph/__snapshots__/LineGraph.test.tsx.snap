// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineGraph should render 1`] = `
<DocumentFragment>
  <div>
    <svg
      class="w-full h-5"
      data-testid="LineGraph__svg"
    >
      <rect
        class="fill-current text-theme-success-600"
        data-testid="LineGraph__item"
        height="8"
        rx="4"
        width="42"
        x="0"
        y="8"
      >
        <animate
          attributeName="height"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="16"
        />
        <animate
          attributeName="height"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="16"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="4"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="8"
          to="4"
        />
        <animate
          attributeName="y"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="0"
        />
        <animate
          attributeName="y"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="0"
          to="8"
        />
      </rect>
      <rect
        class="fill-current text-theme-warning-600"
        data-testid="LineGraph__item"
        height="8"
        rx="4"
        width="22"
        x="50"
        y="8"
      >
        <animate
          attributeName="height"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="16"
        />
        <animate
          attributeName="height"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="16"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="4"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="8"
          to="4"
        />
        <animate
          attributeName="y"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="0"
        />
        <animate
          attributeName="y"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="0"
          to="8"
        />
      </rect>
      <rect
        class="fill-current text-theme-info-600"
        data-testid="LineGraph__item"
        height="8"
        rx="4"
        width="20"
        x="80"
        y="8"
      >
        <animate
          attributeName="height"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="16"
        />
        <animate
          attributeName="height"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="16"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="4"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="8"
          to="4"
        />
        <animate
          attributeName="y"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="0"
        />
        <animate
          attributeName="y"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="0"
          to="8"
        />
      </rect>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`LineGraph should render as empty 1`] = `
<DocumentFragment>
  <div>
    <svg
      class="w-full h-5"
      data-testid="LineGraph__svg"
    >
      <rect
        class="fill-current text-theme-secondary-300 dark:text-theme-secondary-800"
        data-testid="LineGraph__empty"
        height="8"
        rx="4"
        width="100"
        x="0"
        y="8"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`LineGraph should render legend 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-end mb-1"
      data-testid="LineGraph__legend"
    >
      <ul>
        <li>
          item 1 value: 50
        </li>
        <li>
          item 2 value: 30
        </li>
        <li>
          item 3 value: 20
        </li>
      </ul>
    </div>
    <svg
      class="w-full h-5"
      data-testid="LineGraph__svg"
    >
      <rect
        class="fill-current text-theme-success-600"
        data-testid="LineGraph__item"
        height="8"
        rx="4"
        width="42"
        x="0"
        y="8"
      >
        <animate
          attributeName="height"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="16"
        />
        <animate
          attributeName="height"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="16"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="4"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="8"
          to="4"
        />
        <animate
          attributeName="y"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="0"
        />
        <animate
          attributeName="y"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="0"
          to="8"
        />
      </rect>
      <rect
        class="fill-current text-theme-warning-600"
        data-testid="LineGraph__item"
        height="8"
        rx="4"
        width="22"
        x="50"
        y="8"
      >
        <animate
          attributeName="height"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="16"
        />
        <animate
          attributeName="height"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="16"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="4"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="8"
          to="4"
        />
        <animate
          attributeName="y"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="0"
        />
        <animate
          attributeName="y"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="0"
          to="8"
        />
      </rect>
      <rect
        class="fill-current text-theme-info-600"
        data-testid="LineGraph__item"
        height="8"
        rx="4"
        width="20"
        x="80"
        y="8"
      >
        <animate
          attributeName="height"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="16"
        />
        <animate
          attributeName="height"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="16"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="4"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="8"
          to="4"
        />
        <animate
          attributeName="y"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="0"
        />
        <animate
          attributeName="y"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="0"
          to="8"
        />
      </rect>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`LineGraph should render tooltip on hover 1`] = `
<DocumentFragment>
  <div>
    <div
      class="absolute z-10 duration-200 transition-opacity opacity-0 hidden"
      style="top: -48px;"
    >
      <div
        data-testid="TooltipContent"
      >
        item 2 value: 30
      </div>
    </div>
    <svg
      class="w-full h-5"
      data-testid="LineGraph__svg"
    >
      <rect
        class="fill-current text-theme-success-600"
        data-testid="LineGraph__item"
        height="8"
        rx="4"
        width="42"
        x="0"
        y="8"
      >
        <animate
          attributeName="height"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="16"
        />
        <animate
          attributeName="height"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="16"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="4"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="8"
          to="4"
        />
        <animate
          attributeName="y"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="0"
        />
        <animate
          attributeName="y"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="0"
          to="8"
        />
      </rect>
      <rect
        class="fill-current text-theme-warning-600"
        data-testid="LineGraph__item"
        height="8"
        rx="4"
        width="22"
        x="50"
        y="8"
      >
        <animate
          attributeName="height"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="16"
        />
        <animate
          attributeName="height"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="16"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="4"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="8"
          to="4"
        />
        <animate
          attributeName="y"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="0"
        />
        <animate
          attributeName="y"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="0"
          to="8"
        />
      </rect>
      <rect
        class="fill-current text-theme-info-600"
        data-testid="LineGraph__item"
        height="8"
        rx="4"
        width="20"
        x="80"
        y="8"
      >
        <animate
          attributeName="height"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="16"
        />
        <animate
          attributeName="height"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="16"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="4"
          to="8"
        />
        <animate
          attributeName="rx"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="8"
          to="4"
        />
        <animate
          attributeName="y"
          begin="mouseover"
          dur="0.1s"
          fill="freeze"
          from="8"
          to="0"
        />
        <animate
          attributeName="y"
          begin="mouseleave"
          dur="0.15s"
          fill="freeze"
          from="0"
          to="8"
        />
      </rect>
    </svg>
  </div>
</DocumentFragment>
`;
